### ISSUE TITLE
Outcome Always Success?

### ISSUE THREAD

**Issue by skoonin** [8BitJonny/gh-get-current-pr#187]
## Subject of the issue

Clarification on documentation.   I am wondering if I am interpreting it correctly that if the `outcome` of the action is that the commit is NOT part of a PR, it will **not** return `success`? 

In my testing, it is returning `success` regardless of it a PR exists. So using the conditional `if: steps.PR.outcome == 'success'` doesn't really help me in directing things depending on if it is a PR or not. 

I am able to use this how I need it by focusing on the pr number output, however the outcome being successful seems strange to me? 

thanks for creating this! 

## Your git environment
using a self-hosted github runner in github enterprise v3.7 (latest)

## Steps to reproduce
In theory this should work everywhere and not require my environment.  My workflow is below.

## Expected behaviour
If commit is not part of a PR, action will not return success

## Actual behaviour
success is returned regardless.  However the PR information outputs are empty.

Workflow: 
```
name: Workflow

on:
  push:
    branches-ignore:
      - main

# stops current jobs if a new commit is pushed to the branch
concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

PR:
    runs-on: [dev, terraform]
    outputs:
      PR_STATUS: ${{ steps.pr-check.outputs.PR_STATUS }}
      PR_NUMBER: ${{ steps.PR.outputs.number }}
      PR_OUTCOME: ${{ steps.PR.outcome }}
    steps:
      - uses: synced-actions/gh-get-current-pr@2.1.3
        id: PR
      - id: pr-check
        run: |
          if [ ${{ steps.PR.outputs.number }} == 0 ]; then
            echo "PR_STATUS=false" >> $GITHUB_OUTPUT
          else
            echo "PR_STATUS=true" >> $GITHUB_OUTPUT
          fi

  test:
    needs: PR
    runs-on: [dev, terraform]
    steps:
      - run: |
          echo "Status:" ${{ needs.PR.outputs.PR_STATUS }} and Number: ${{ needs.PR.outputs.PR_NUMBER }}""
      - if: needs.PR.outputs.PR_STATUS == true
        run: |
          echo "TRUE PR status is ${{ needs.PR.outputs.PR_STATUS }}"
      - if: needs.PR.outputs.PR_STATUS != true
        run: |
          echo "FALSE PR status : ${{ needs.PR.outputs.PR_STATUS }}"
      - if: needs.PR.outputs.PR_NUMBER == 0
        run: |
          echo "empty PR NUMBER : ${{ needs.PR.outputs.PR_NUMBER }}"
      - if: needs.PR.outputs.PR_NUMBER != 0
        run: |
          echo "existing PR NUMBER is ${{ needs.PR.outputs.PR_NUMBER }}"
```

**Comment by 8BitJonny** [comment]
Thanks for this detailed error message! Really appreciate it.

Yes, you are very correct. This is indeed very confusing, that even with no found PR `outcome` is always success.

As for now your approach already does the trick in terms of workaround. I even believe you can reduce the if condition down to 
```
- if: steps.PR.outputs.PR
```
or any other of the outputs. As none of them are set at all, when no PR is found a simple `if output.pr_number` should do the trick and doesn't need a `if output.pr_number != 0`

**Comment by 8BitJonny** [comment]
As for the next steps: This is very easy to fix. The action just needs to return a non zero exit code, when no PR is found. Then the `outcome` can be used again.
I'll make the bug fix version ready over the weekend and ping here once its online

**Comment by 8BitJonny** [comment]
Alright, as you can see the PR is ready BUT while setting the PR up and adding test cases on how everything looks and feels like when working with a job and a non zero exit code, I want to have a quick chat here.
I have noticed three points which speak against this behaviour.

1.) Breaking Change
This is a breaking change. Even if this is a breaking change towards, what one might argue a better syntax, it is still a breaking change.

2.) Mixing two different levels of Errors
Until now, you get two different levels of errors shown to the user.
You have the case where the action succeeds in running and just doesn't find a PR where in the current case 
```
- if: steps.PR.outputs.PR
```
would catch these instances.
And then you have the instances where the action itself is failing and throwing an error. These issues might be network related, source code related, ...

If we know change the behaviour you don't make this separation anymore. It might not be needed, but I kinda like it.

3.) Github's Error handling
And maybe the biggest reason that threw me off, is how Github handles errors when running multiple steps.
Because by default all following steps would just be skipped.
You actively also would have to set `continue-on-error` to true for this job ([docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error)) in order for the following steps still to execute.
Or you'd have to add `if: always()` to all the following jobs

**Summary**
All these three points together make me not too sure anymore if this change is really worth it.
Because currently:
- the github error handling kicks only in if this job REALLY fails. So in 99% of the cases you don't have to think about their error handling
- you can still check for success or error of this job with `if: steps.PR.outputs.PR`
- or even handle a full error with `if: always() && steps.PR.outcome == 'success'` or `if: always() && steps.PR.outcome == 'failure'`

But yeah, in this case the README would have to be adjusted because it is currently wrong for my described error handling

**Comment by 8BitJonny** [comment]
Feedback on my points, highly welcome otherwise I'll close this issue and fix the Readme in a couple of days.
@skoonin

**Comment by skoonin** [comment]
@8BitJonny yea i understand your thoughts on the confusion btwn errors.. maybe it would make more sense instead to create a new output that is just for if the PR exists or not?  `steps.PR.pr-exists` or something like that? I was just using PR number since it solved my problem, but if you want this functionality and don't want to change the outcome behaviour this may make sense?

In this case, you should only need to update the docs to no longer use the `outcome` status at all and instead use the exists output.

(Just FYI, I haven't coded actions myself yet, but I am assuming `outcome` is a standard Actions return?)

**Comment by 8BitJonny** [comment]
I really like this idea. I'll go ahead and implement a new `pr-found` output property, which is always set and either is true when a PR is found or false when no PR is found. Docs will be updated too.
I like `pr-found` more than `pr-exists` because a pr might very well exist but might not be returned due to the configured filtering of `filterDraft` & `filterClosed` PRs

**Comment by skoonin** [comment]
thanks for the clarification and yea `pr-found` is a better way to go.  Thanks for the quick response for this @8BitJonny!

**Comment by 8BitJonny** [comment]
Haven't forgot you. Just had a full weekend 😅  Hope to get it done in the next days.

**Comment by skoonin** [comment]
no rush man!
On Dec 5, 2022 at 4:30 PM -0800, Timon Christiansen ***@***.***>, wrote:

[EXTERNAL] Use caution when opening attachments or links.


Haven't forgot you. Just had a full weekend 😅 Hope to get it done in the next days.

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2F8BitJonny%2Fgh-get-current-pr%2Fissues%2F187%23issuecomment-1338486010&data=05%7C01%7Cskoonin%40corescientific.com%7C92ae18c11d06445f094908dad7211ec2%7C7fd0b7127cce4cc0a64ac86fd9fcfd80%7C1%7C0%7C638058834504767142%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pxOqzmtpmYfb5wUe4JYUKlYe6jywOhwaYk%2Fpph9Os74%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAC5VJKCBXTFSIK2LFV5KZ53WL2CLPANCNFSM6AAAAAASJU3M6A&data=05%7C01%7Cskoonin%40corescientific.com%7C92ae18c11d06445f094908dad7211ec2%7C7fd0b7127cce4cc0a64ac86fd9fcfd80%7C1%7C0%7C638058834504767142%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=iZeEi569X2C%2FsGEiv1mF7ayRQ6W7wSZ%2BR2sAAmDVDI8%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

The information contained in this email message is intended only for use of the individual or entity named above. If the reader of this message is not the intended recipient, or the employee or agent responsible to deliver it to the intended recipient, you are hereby notified that any dissemination, distribution or copying of this communication is strictly prohibited. If you have received this communication in error, please immediately notify us by email noted above and destroy the original message.

**Comment by 8BitJonny** [comment]
released under v2.2.0 just now

**Comment by 8BitJonny** [comment]
thanks for your help and input
